GenericList/list.h

```diff
@@ -0,0 +1,260 @@
+#pragma once
+
+#include <iostream>
+#include <stdexcept>
+#include <string>
+#include <memory>
+#include <cassert>
+#include <functional>
+#include <algorithm>
+#include <type_traits>
+
+// Never use header-wide using directives ("using namespace") in the header!!
+// Explanation: https://stackoverflow.com/questions/5849457/using-namespace-in-c-headers
+// Basically, when this header is included, the "using namespace" will be included too, possibly causing un-intended clashes in variable/function/etc.. namess
+// Note: as long as code doesn't #include a .cpp file, it's fine to use using directives in that source file
+
+//Generic type, allow for stateful Allocator if user desires it
+template <typename T, typename Allocator = std::allocator<T>>
+class List {
+    static_assert(!std::is_void_v<T>, "void type is not allowed");
+    static_assert(!std::is_reference_v<T>, "reference type is not allowed");
+    //C++ Standard �8.3.2/4: There shall be no references to references, no arrays of references, and no pointers to references.
+
+
+    //note: elements are value copies of original objects (copy-by-value, not copy-by-reference)
+public:
+    List() : dataAllocator(), data(nullptr), capacity(0), count(0) {
```

Comment:
I forgot to mention, if you haven't explicitly instantiated a class template, function members of class templates don't really "exist". The compiler will check the C++ syntax on anything that isn't dependent on the template parameters but until you actually use it, they don't exist. So suppose you have an allocator that isn't default constructible:
```c++
struct my_alloc {
    my_alloc() = delete;
};
```

`List<int, my_alloc>` should still compile because `List<int, my_alloc>()` doesn't exist until you use it. See [example](https://godbolt.org/z/s18ersbcz)

Reply:
Nah, it's just FYI, there's nothing wrong here, it's supposed to be this way. It's very rare that you need to account for the explicit instantiation. Aside from the ones I've put in, I've never seen it in Unreal. 

If you really wanna take it into account, you can add another layer:
```c++
template<typename T>
struct data_container {
    T* data = nullptr;
    size_t capacity = 0;
    size_t count = 0;
};
template<typename T, typename Alloc>
class List : private data_container {
    constexpr List() noexcept = default;
private:
    Alloc alloc;
};
``` 
In this case, if `Alloc()` is deleted so will `List()` .